File: frontend/src/utils/salaryCalculator.js
import { insuranceCalculator, taxCalculator } from './insuranceCalculator';
import { isHoliday } from 'holiday-kr';
import { format } from 'date-fns';

export class SalaryCalculator {
  constructor(startDate, endDate) {
    this.startDate = new Date(startDate);
    this.endDate = new Date(endDate);
    this.REGULAR_WORK_HOURS_PER_DAY = 8;  // 1일 소정근로시간
    this.MONTHLY_WORK_HOURS = 209;  // 월 소정근로시간
    this.NIGHT_START = '22:00';  // 야간근로 시작
    this.NIGHT_END = '06:00';    // 야간근로 종료
  }

  calculateBasicPay(annualSalary) {
    const monthlyBasePay = Math.floor(annualSalary / 12);
    const dailyBasePay = Math.floor(monthlyBasePay / 30);
    
    // 선택된 일수 계산
    const selectedDays = Math.ceil((this.endDate - this.startDate) / (1000 * 60 * 60 * 24)) + 1;
    
    return dailyBasePay * selectedDays;
  }

  calculateRegularWage(annualSalary) {
    const monthlyBasePay = Math.floor(annualSalary / 12);
    return Math.floor(monthlyBasePay / this.MONTHLY_WORK_HOURS);
  }

  calculateWorkHours(checkIn, checkOut) {
    // 근무시간 계산 (휴게시간 자동 차감)
    const start = new Date(checkIn);
    const end = new Date(checkOut);
    
    // 총 근무시간 계산 (시간 단위)
    const totalHours = (end - start) / (1000 * 60 * 60);

    // 휴게시간 차감
    return this.adjustBreakTime(totalHours);
  }

  adjustBreakTime(workHours) {
    if (workHours >= 8) {
      return workHours - 1;  // 1시간 휴게
    } else if (workHours >= 4) {
      return workHours - 0.5;  // 30분 휴게
    }
    return workHours;
  }

  isHoliday(dateStr) {
    const date = new Date(dateStr);
    const day = date.getDay();
    
    // 공휴일 체크 (패키지의 isHoliday 함수 사용)
    const isPublicHoliday = isHoliday(date);
    
    return day === 0 || day === 6 || isPublicHoliday;
  }

  calculateTotalPay(attendanceData, employee) {
    const regularWage = this.calculateRegularWage(employee.base_salary);
    
    const overtimePay = this.calculateOvertimePay(attendanceData, regularWage);
    const nightShiftPay = this.calculateNightShiftPay(attendanceData, regularWage);
    const holidayPay = this.calculateHolidayPay(attendanceData, regularWage);

    const totalPay = this.calculateBasicPay(employee.base_salary) + overtimePay + nightShiftPay + holidayPay;
    
    // 4대보험 및 소득세 계산
    const deductions = insuranceCalculator.calculateAll(
      totalPay, 
      employee.dependents, 
      employee.tax_deductible_children
    );

    // 세금 계산
    const taxes = taxCalculator.calculateTax(
      totalPay,
      employee.dependents,
      employee.tax_deductible_children
    );

    return {
      regularWage,
      overtimePay,
      nightShiftPay,
      holidayPay,
      totalPay,
      ...deductions,
      ...taxes
    };
  }

  calculateOvertimePay(attendanceData, regularWage) {
    let totalOvertimeHours = 0;

    attendanceData.forEach(record => {
      if (!['정상', '지각'].includes(record.attendance_type)) return;

      const start = new Date(record.check_in);
      const end = new Date(record.check_out);
      const dayEnd = new Date(start);
      dayEnd.setHours(18, 0, 0);

      if (end > dayEnd) {
        const overtime = (end - dayEnd) / (1000 * 60 * 60);
        totalOvertimeHours += this.adjustBreakTime(overtime);
      }
    });

    return Math.floor(totalOvertimeHours * regularWage * 1.5);
  }

  calculateNightShiftPay(attendanceData, regularWage) {
    let totalNightHours = 0;

    attendanceData.forEach(record => {
      if (!['정상', '지각'].includes(record.attendance_type)) return;

      const start = new Date(record.check_in);
      const end = new Date(record.check_out);
      
      // 야간근로시간 계산 (22:00-06:00)
      if (start.getHours() >= 22 || end.getHours() < 6) {
        const nightHours = this.calculateNightHours(start, end);
        totalNightHours += nightHours;
      }
    });

    return Math.floor(totalNightHours * regularWage * 0.5);
  }

  calculateHolidayPay(attendanceData, regularWage) {
    let totalHolidayPay = 0;
    const groupedByDate = this.groupByDate(attendanceData);

    Object.entries(groupedByDate).forEach(([date, records]) => {
      // 휴일근무 or 공휴일 근무 체크
      const isHolidayWork = this.isHoliday(date) || 
        records.some(r => r.attendance_type === '휴일근무');

      if (isHolidayWork) {
        let dailyHours = records.reduce((sum, record) => {
          if (!['정상', '지각', '휴일근무'].includes(record.attendance_type)) return sum;
          return sum + this.calculateWorkHours(record.check_in, record.check_out);
        }, 0);

        // 8시간 초과분 계산
        const regularHours = Math.min(dailyHours, 8);
        const overtimeHours = Math.max(dailyHours - 8, 0);

        totalHolidayPay += Math.floor(regularHours * regularWage * 1.5) +
                          Math.floor(overtimeHours * regularWage * 2);
      }
    });

    return totalHolidayPay;
  }

  calculateNightHours(start, end) {
    // 야간근로시간 계산 헬퍼 함수
    const nightStart = new Date(start);
    nightStart.setHours(22, 0, 0);
    const nightEnd = new Date(start);
    nightEnd.setHours(29, 0, 0); // 다음날 06:00를 29:00으로 표현

    const workStart = start.getHours() >= 22 ? start : nightStart;
    const workEnd = end.getHours() < 6 ? end : nightEnd;

    return Math.max(0, (workEnd - workStart) / (1000 * 60 * 60));
  }

  groupByDate(attendanceData) {
    return attendanceData.reduce((acc, record) => {
      const date = record.check_in.split(' ')[0];
      if (!acc[date]) acc[date] = [];
      acc[date].push(record);
      return acc;
    }, {});
  }
}

File: backend/utils/salary_calculator.py
from datetime import datetime
from .insurance_calculator import InsuranceCalculator

class SalaryCalculator:
    def __init__(self, start_date, end_date):
        self.start_date = datetime.strptime(start_date, '%Y-%m-%d')
        self.end_date = datetime.strptime(end_date, '%Y-%m-%d')
        self.REGULAR_WORK_HOURS_PER_DAY = 8  # 1일 소정근로시간
        self.MONTHLY_WORK_HOURS = 209  # 월 소정근로시간
        self.NIGHT_START = '22:00'  # 야간근로 시작
        self.NIGHT_END = '06:00'    # 야간근로 종료

    def calculate_basic_pay(self, annual_salary):
        monthly_base_pay = annual_salary // 12
        daily_base_pay = monthly_base_pay // 30
        selected_days = ((self.end_date - self.start_date).days + 1)
        return daily_base_pay * selected_days

    def calculate_regular_wage(self, annual_salary):
        monthly_base_pay = annual_salary // 12
        return monthly_base_pay // self.MONTHLY_WORK_HOURS

    def calculate_work_hours(self, check_in, check_out):
        start = datetime.strptime(check_in, '%Y-%m-%dT%H:%M:%S')
        end = datetime.strptime(check_out, '%Y-%m-%dT%H:%M:%S')
        total_hours = (end - start).total_seconds() / 3600
        return self.adjust_break_time(total_hours)

    def adjust_break_time(self, work_hours):
        if work_hours >= 8:
            return work_hours - 1
        elif work_hours >= 4:
            return work_hours - 0.5
        return work_hours

    def is_holiday(self, date_str):
        date = datetime.strptime(date_str, '%Y-%m-%d')
        day = date.weekday()  # 0=월요일, 6=일요일
        return day == 5 or day == 6  # 토요일 or 일요일

    def calculate_overtime_pay(self, attendance_data, regular_wage):
        total_overtime_hours = 0
        for record in attendance_data:
            if record['attendance_type'] not in ['정상', '지각']:
                continue
            start = datetime.strptime(record['check_in'], '%Y-%m-%dT%H:%M:%S')
            end = datetime.strptime(record['check_out'], '%Y-%m-%dT%H:%M:%S')
            day_end = start.replace(hour=18, minute=0, second=0)
            if end > day_end:
                overtime = (end - day_end).total_seconds() / 3600
                total_overtime_hours += self.adjust_break_time(overtime)
        return int(total_overtime_hours * regular_wage * 1.5)

    def calculate_night_shift_pay(self, attendance_data, regular_wage):
        total_night_hours = 0
        for record in attendance_data:
            if record['attendance_type'] not in ['정상', '지각']:
                continue
            start = datetime.strptime(record['check_in'], '%Y-%m-%dT%H:%M:%S')
            end = datetime.strptime(record['check_out'], '%Y-%m-%dT%H:%M:%S')
            if start.hour >= 22 or end.hour < 6:
                night_hours = self.calculate_night_hours(start, end)
                total_night_hours += night_hours
        return int(total_night_hours * regular_wage * 0.5)

    def calculate_holiday_pay(self, attendance_data, regular_wage):
        total_holiday_pay = 0
        grouped_by_date = self.group_by_date(attendance_data)
        for date, records in grouped_by_date.items():
            is_holiday_work = self.is_holiday(date) or any(r['attendance_type'] == '휴일근무' for r in records)
            if is_holiday_work:
                daily_hours = sum(
                    self.calculate_work_hours(r['check_in'], r['check_out'])
                    for r in records if r['attendance_type'] in ['정상', '지각', '휴일근무']
                )
                regular_hours = min(daily_hours, 8)
                overtime_hours = max(daily_hours - 8, 0)
                total_holiday_pay += int(regular_hours * regular_wage * 1.5) + int(overtime_hours * regular_wage * 2)
        return total_holiday_pay

    def calculate_night_hours(self, start, end):
        night_start = start.replace(hour=22, minute=0, second=0)
        night_end = start.replace(hour=29, minute=0, second=0)  # 다음날 06:00
        work_start = start if start.hour >= 22 else night_start
        work_end = end if end.hour < 6 else night_end
        return max(0, (work_end - work_start).total_seconds() / 3600)

    def group_by_date(self, attendance_data):
        groups = {}
        for record in attendance_data:
            date = record['check_in'].split('T')[0]
            if date not in groups:
                groups[date] = []
            groups[date].append(record)
        return groups